design/client-initiated-backchannel-authentication-flow.md

```diff
@@ -0,0 +1,494 @@
+
+* interval : supported
+
+If this "interval" is set to 0, the token request throttling is deactivated.
+
+According to CIBA protocol specification, the access not waiting this interval is incurred the penalty of 5 sec as an additional wait.
+
+#### Authorization
+
+Authorization by AD works the same as the current keycloak's authorization code flow. If the CD's "Consent Required" setting is "Yes", the following scope value are presented to the end-user on the AD's UI:
+
+* scope                : specified by Backchannel Authentication Request's "scope" parameter
+* client default scope : set by keycloak Admin Console's Client Settings
+
+#### Binding Message
+
+On CIBA Protocol specification Section [7.1. Authentication Request](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#auth_request), Binding Message is supported.
+
+#### Pairwise Identifiers
+
+On CIBA Protocol specification Section [4. Registration and Discovery Metadata](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#registration), Poll Modes with Pairwise Identifiers is NOT supported.
+
+#### User Code
+
+On CIBA Protocol specification Section [7.1. Authentication Request](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#auth_request), User Code is NOT supported.
+
+#### Requested Expiry
+
+On CIBA Protocol specification Section [7.1. Authentication Request](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#auth_request), Requested Expiry is NOT supported.
+
+#### Signed Authentication Request
+
+On CIBA Protocol specification Section [7.1.1. Signed Authentication Request](https://openid.net/specs/openid-client-initiated-backchannel-authentication-core-1_0.html#signed_auth_request), Signed Authentication Request is NOT supported.
+
+### Outside CIBA Protocol
+
+The process of AuthN/AuthZ by AD is highly use-cases dependent (e.g. Push notification to the end-user's smart device, SMS message to the end-user's smart device, etc.).
+
+To keep the design in use-cases independent, this design document describes the interface between inside and outside CIBA protocol.
+
+However, if keycloak lacks the actual implementation of these interfaces, the user of keycloak does not use CIBA flow unless they implement these interfaces by themselves. It is inconvenient for users.
+
+Therefore, this design document also defines the default implementation for these interfaces.
+
+### Runtime Environment
+
+This version of the design document considers both standalone and clustering environments but not consider multi-cluster environment.
+
+### Settings
+
+To configure inside CIBA part's operation, CIBA policy is introduced as part of Admin Console's Realm Settings.
```

Comment:
I'm not sure about having this as a realm setting.

1) Clients may have different needs. Although we only support `poll`, later we may support other modes.
2) `expires_in` is OK as a default for all clients within a realm. But later we may want to also allow clients to send this.
3) Clients may have different requirements for `interval`.

In addition to that, how we enable CIBA for a client ? Do we want a switch similar to what we have for `Direct Grant` ?

Reply:
I see, I'd like to enable the settings per client but also support per realm as default setting (setting per client supersede the one per realm). WDYT?

As commented earlier, we can control which client be allowed to use CIBA flow using [Client Policies ](https://github.com/keycloak/keycloak-community/blob/master/design/client-policies.md) . I'd like to add this description about it on this PR's design document. WDYT?

Reply:
Looks good to me to have a default realm setting. It should be easier to configure if all your clients share the same configuration (we don't have the concept of client templates anymore).

As I mentioned before, not sure about client policies for such a thing.

Reply:
Therefore, I'll revise this PR's design document to write about this settings (setting per client supersede the one per realm).

For example, by using Client Policies, keycloak can accept the bachchannel authentication request sent from the clients whose client roles include "ciba-flow-allowed" and refuse it from the clients whose client roles does not include "ciba-flow-allowed".

Reply:
Sounds good.